File: API -Windows/csharp/spi_eeprom.cs
/*=========================================================================
| (c) 2015 Total Phase, Inc.
|--------------------------------------------------------------------------
| Project : Promira Sample Code
| File    : spi_eeprom.cs
|--------------------------------------------------------------------------
| Perform simple read and write operations to an SPI EEPROM device.
|--------------------------------------------------------------------------
| Redistribution and use of this file in source and binary forms, with
| or without modification, are permitted.
|
| THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
| "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
| LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
| FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE
| COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
| INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
| BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
| LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
| CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
| LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
| ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
| POSSIBILITY OF SUCH DAMAGE.
 ========================================================================*/

using System;
using TotalPhase;


/*=========================================================================
| CLASS
 ========================================================================*/
public class PMSpiEeprom {

    /*=====================================================================
    | CONSTANTS
     ====================================================================*/
    public const int PAGE_SIZE = 32;
    public const int SS_MASK   = 0x1;


    /*=====================================================================
    | STATIC FUNCTIONS (APP)
     ====================================================================*/
    private const string APP_NAME = "com.totalphase.promact_is";

    private static int DevOpen (string ip,
                                ref int pm,
                                ref int conn,
                                ref int channel)
    {
        int ret;
        pm = PromiraApi.pm_open(ip);
        if (pm <= 0) {
            Console.WriteLine("Unable to open Promira platform on {0}\n", ip);
            Console.WriteLine("Error code = {0}\n", pm);
            return -1;
        }
        ret = PromiraApi.pm_load(pm, APP_NAME);
        if (ret < 0) {
            Console.WriteLine("Unable to load the application({0})\n",
                              APP_NAME);
            Console.WriteLine("Error code = {0}\n", ret);
            return -1;
        }

        conn = Promact_isApi.ps_app_connect(ip);
        if (conn <= 0) {
            Console.WriteLine("Unable to open the application on {0}\n", ip);
            Console.WriteLine("Error code = {0}\n", conn);
            return -1;
        }

        channel = Promact_isApi.ps_channel_open(conn);
        if (channel <= 0) {
            Console.WriteLine("Unable to open the channel on {0}\n", ip);
            Console.WriteLine("Error code = {0}\n", channel);
            return -1;
        }

        return 0;
    }

    private static void DevClose (int pm, int conn, int channel)
    {
        Promact_isApi.ps_channel_close(channel);
        Promact_isApi.ps_app_disconnect(conn);
        PromiraApi.pm_close(pm);
    }

    private static int DevCollect (int collect, byte[] buf)
    {
        int ret = 0, offset = 0;
        int t, length = 0, result = 0;
        if (collect < 0) {
            Console.WriteLine(
                "{0}\n", Promact_isApi.ps_app_status_string(collect));
            return collect;
        }
        while (true) {
            t = Promact_isApi.ps_collect_resp(collect, ref length,
                                              ref result, -1);
            if (t == (int)PromiraAppStatus.PS_APP_NO_MORE_CMDS_TO_COLLECT)
                break;
            else if (t < 0) {
                Console.WriteLine(
                    "{0}\n", Promact_isApi.ps_app_status_string(t));
                return t;
            }
            if (t == (int)PromiraSpiCommand.PS_SPI_CMD_READ) {
                if (buf.Length > offset && buf.Length != 0) {
                    byte[] dataIn = new byte[length];
                    byte word_size = 0;
                    ret = Promact_isApi.ps_collect_spi_read(
                        collect, ref word_size,
                        (uint)(buf.Length - offset), dataIn);

                    Array.Copy(dataIn, 0, buf, offset, ret);
                }

                offset += ret;
            }
        }
        return offset;
    }

    private static void SpiMasterOE (int channel, int queue, byte enable)
    {
        int collect;
        byte queue_type = 0;
        Promact_isApi.ps_queue_clear(queue);
        Promact_isApi.ps_queue_spi_oe(queue, enable);
        collect = Promact_isApi.ps_queue_submit(
            queue, channel, 0, ref queue_type);
        DevCollect(collect, new byte[0]);
    }


    /*=====================================================================
    | FUNCTIONS
     ====================================================================*/
    static void _writeMemory (int channel, int queue,
                              short addr, short length, bool zero)
    {
        short i, n;
        int collect;
        byte queue_type = 0;

        byte[] dataOut = null;
        byte[] writeEnableBuf = new byte[1];

        // Write to the SPI EEPROM
        //
        // The AT25080A EEPROM has 32 byte pages.  Data can written
        // in pages, to reduce the number of overall SPI transactions
        // executed through the Aardvark adapter.
        n = 0;
        while (n < length) {
            // Clear the queue
            Promact_isApi.ps_queue_clear(queue);

            // Calculate the amount of data to be written this iteration
            // and make sure dataOut is just large enough for it.
            int size = Math.Min(((addr & (PAGE_SIZE - 1)) ^ (PAGE_SIZE - 1)) + 1,
                                length - n);
            byte[] data = new byte[size];

            if (dataOut == null)
                dataOut = new byte[3];

            // Send write enable command
            writeEnableBuf[0] = 0x06;
            Promact_isApi.ps_queue_spi_ss(queue, SS_MASK);
            Promact_isApi.ps_queue_spi_write(queue, 0, 8,
                                             1, writeEnableBuf);
            Promact_isApi.ps_queue_spi_ss(queue, 0);

            // Assemble write command and address
            dataOut[0] = 0x02;
            dataOut[1] = (byte)(((addr + n) >> 8) & 0xff);
            dataOut[2] = (byte)(((addr + n) >> 0) & 0xff);

            // Assemble a page of data
            i = 0;
            do {
                data[i++] = zero ? (byte)0 : (byte) n;
                ++n;
            } while (n < length && ((addr + n) & (PAGE_SIZE - 1)) != 0);

            // Write the transaction
            Promact_isApi.ps_queue_spi_ss(queue, SS_MASK);
            Promact_isApi.ps_queue_spi_write(queue, 0, 8,
                                             (uint)dataOut.Length, dataOut);
            Promact_isApi.ps_queue_spi_write(queue, 0, 8,
                                             (uint)data.Length, data);
            Promact_isApi.ps_queue_spi_ss(queue, 0);
            Promact_isApi.ps_queue_spi_delay_ns(queue, 10 * 1000 * 1000);

            collect = Promact_isApi.ps_queue_submit(queue, channel, 0,
                                                    ref queue_type);
            DevCollect(collect, new byte[0]);
        }
    }

    static void _readMemory (int channel, int queue, short addr, short length)
    {
        int count;
        int i;
        int collect;
        byte queue_type = 0;

        byte[] dataOut = new byte[3];
        byte[] dataIn  = new byte[length + 3];

        // Clear the queue
        Promact_isApi.ps_queue_clear(queue);

        // Assemble read command and address
        dataOut[0] = 0x03;
        dataOut[1] = (byte)((addr >> 8) & 0xff);
        dataOut[2] = (byte)((addr >> 0) & 0xff);

        // Write length+3 bytes for data plus command and 2 address bytes
        Promact_isApi.ps_queue_spi_ss(queue, SS_MASK);
        Promact_isApi.ps_queue_spi_write(queue, (PromiraSpiIOMode)0, 8,
                                         (uint)dataOut.Length, dataOut);
        Promact_isApi.ps_queue_spi_write_word(queue, 0, 8,
                                              (uint)length, 0);
        Promact_isApi.ps_queue_spi_ss(queue, 0);

        collect = Promact_isApi.ps_queue_submit(queue, channel, 0,
                                                ref queue_type);
        count = DevCollect(collect, dataIn);

        if (count < 0) {
            Console.WriteLine("error: {0}", count);
        }
        else if (count != length + 3) {
            Console.WriteLine("error: read {0} bytes (expected {1})",
                              count - 3, length);
        }

        // Dump the data to the screen
        Console.Write("\nData read from device:");
        for (i = 0; i < length; ++i) {
            if ((i & 0x0f) == 0)
                Console.Write("\n{0:x4}:  ", addr + i);
            Console.Write("{0:x2} ", dataIn[i + 3] & 0xff);
            if (((i + 1) & 0x07) == 0)
                Console.Write(" ");
        }
        Console.WriteLine();
    }


    /*=====================================================================
    | MAIN PROGRAM
     ====================================================================*/
    public static void Main (string[] args) {
        int pm      = 0;
        int conn    = 0;
        int channel = 0;
        int queue   = 0;

        string ip      = "";
        int    bitrate = 100;
        int    mode    = 0;
        short  addr;
        short  length;
        int    res;

        if (args.Length != 6) {
            Console.WriteLine(
                "usage: spi_eeprom IP BITRATE read  MODE ADDR LENGTH");
            Console.WriteLine(
                "usage: spi_eeprom IP BITRATE write MODE ADDR LENGTH");
            Console.WriteLine(
                "usage: spi_eeprom IP BITRATE zero  MODE ADDR LENGTH");
            return;
        }

        // Parse the IP argument
        try {
            ip = Convert.ToString(args[0]);
        }
        catch (Exception) {
            Console.WriteLine("Error: invalid IP address");
            return;
        }

        // Parse the bitrate argument
        try {
            bitrate = Convert.ToInt32(args[1]);
        }
        catch (Exception) {
            Console.WriteLine("Error: invalid bitrate");
            return;
        }

        string command = args[2];

        // Parse the mode argument
        try {
            mode = Convert.ToInt32(args[3]) & 0x3;
        }
        catch (Exception) {
            Console.WriteLine("Error: invalid mode");
            return;
        }

        // Parse the memory offset argument
        try {
            if (args[4].StartsWith("0x"))
                addr = Convert.ToByte(args[4].Substring(2), 16);
            else
                addr = Convert.ToByte(args[4]);
        }
        catch (Exception) {
            Console.WriteLine("Error: invalid memory addr");
            return;
        }

        // Parse the length
        try {
            length = Convert.ToInt16(args[5]);
        }
        catch (Exception) {
            Console.WriteLine("Error: invalid length");
            return;
        }

        if (mode == 1 || mode == 2) {
            Console.WriteLine(
                "error: spi modes 1 and 2 are not supported by the AT25080A");
            return;
        }

        // Open the device
        res = DevOpen(ip, ref pm, ref conn, ref channel);
        if (res < 0)
            return;

        // Ensure that the SPI subsystem is enabled.
        Promact_isApi.ps_app_configure(channel,
                                       Promact_isApi.PS_APP_CONFIG_SPI |
                                       Promact_isApi.PS_APP_CONFIG_I2C);

        // Power the board using the Promira adapter's power supply.
        Promact_isApi.ps_phy_target_power(
            channel, Promact_isApi.PS_PHY_TARGET_POWER_BOTH);

        // Setup the clock phase
        Promact_isApi.ps_spi_configure(channel,
                                       PromiraSpiMode.PS_SPI_MODE_0,
                                       PromiraSpiBitorder.PS_SPI_BITORDER_MSB,
                                       0);

        // Configure SS
        Promact_isApi.ps_spi_enable_ss(channel, SS_MASK);

        // Set the bitrate
        bitrate = Promact_isApi.ps_spi_bitrate(channel, (ushort)bitrate);
        Console.WriteLine("Bitrate set to {0} kHz", bitrate);

        // Create a queue for SPI transactions
        queue = Promact_isApi.ps_queue_create(
            conn, Promact_isApi.PS_MODULE_ID_SPI_ACTIVE);

        // Enable master output
        SpiMasterOE(channel, queue, 1);

        // Perform the operation
        if (command == "write") {
            _writeMemory(channel, queue, addr, length, false);
            Console.WriteLine("Wrote to EEPROM");
        }
        else if (command == "read") {
            _readMemory(channel, queue, addr, length);
        }
        else if (command == "zero") {
            _writeMemory(channel, queue, addr, length, true);
            Console.WriteLine("Zeroed EEPROM");
        }
        else {
            Console.WriteLine("unknown command: {0}", command);
        }

        // Disable master output
        SpiMasterOE(channel, queue, 0);

        // Destroy the queue
        Promact_isApi.ps_queue_destroy(queue);

        // Close the device and exit
        DevClose(pm, conn, channel);
    }
}

File: promact_is-api-windows-x86_64-v1.60/vb.net/spi_eeprom.vb
'==========================================================================
' (c) 2015  Total Phase, Inc.
'--------------------------------------------------------------------------
' Project : Promira Sample Code
' File    : spi_eeprom.vb
'--------------------------------------------------------------------------
' Perform simple read and write operations to an SPI EEPROM device.
'--------------------------------------------------------------------------
' Redistribution and use of this file in source and binary forms, with
' or without modification, are permitted.
'
' THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
' "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
' LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
' FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE
' COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
' INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
' BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
' LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
' CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
' LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
' ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
' POSSIBILITY OF SUCH DAMAGE.
'==========================================================================
Imports TotalPhase

Module spi_eeprom

    '==========================================================================
    ' CONSTANTS
    '==========================================================================
    Const APP_NAME As String = "com.totalphase.promact_is"
    Const SPI_BITRATE As Integer = 1000  'kHz


    '==========================================================================
    ' FUNCTIONS
    '==========================================================================
    Private Function find_device_ip() As String
        Dim num As Long
        Dim ips(15) As UInt32
        Dim uniqueIds(15) As UInt32
        Dim statuses(15) As UInt32

        ' Find all the attached devices
        num = PromiraApi.pm_find_devices_ext(16, ips, 16, uniqueIds,
                                             16, statuses)

        If num > 0 Then
            Dim ip As UInt32
            Dim i As Long

            ' Print the information on each device
            For i = 0 To num - 1
                ip = ips(i)

                ' Determine if the device is in-use
                If Not (statuses(i) And PromiraApi.PM_DEVICE_NOT_FREE) Then
                    Return (ip And &HFF) & "." & ((ip >> 8) And &HFF) & "." &
                           ((ip >> 16) And &HFF) & "." & ((ip >> 24) And &HFF)
                End If
            Next
        End If
        Return ""
    End Function

    Private Function dev_open(ip As String, ByRef pm As Integer,
                  ByRef conn As Integer, ByRef channel As Integer) As Integer
        Dim ret As Integer
        pm = PromiraApi.pm_open(ip)
        If pm <= 0 Then
            Console.WriteLine("Unable to open Promira platform on " & ip & "\n")
            Console.WriteLine("Error code = " & pm & "\n")
            Return -1
        End If
        ret = PromiraApi.pm_load(pm, APP_NAME)
        If ret < 0 Then
            Console.WriteLine("Unable to load the application(" &
                              APP_NAME & ")\n")
            Console.WriteLine("Error code = " & ret & "\n")
            Return -1
        End If

        conn = Promact_isApi.ps_app_connect(ip)
        If conn <= 0 Then
            Console.WriteLine("Unable to open the application on " & ip & "\n")
            Console.WriteLine("Error code = " & conn & "\n")
            Return -1
        End If

        channel = Promact_isApi.ps_channel_open(conn)
        If channel <= 0 Then
            Console.WriteLine("Unable to open the channel on " & ip & "\n")
            Console.WriteLine("Error code = " & channel & "\n")
            Return -1
        End If

        Return 0
    End Function

    Private Sub dev_close(pm As Integer, conn As Integer, channel As Integer)
        Promact_isApi.ps_channel_close(channel)
        Promact_isApi.ps_app_disconnect(conn)
        PromiraApi.pm_close(pm)
    End Sub

    Private Function dev_collect(collect As Integer, buf() As Byte) As Integer
        Dim ret As Integer
        Dim offset As Integer
        Dim t As Integer
        Dim length As Integer
        Dim result As Integer
        Dim word_size As Byte

        offset = 0

        If (collect < 0) Then
            Console.WriteLine(Promact_isApi.ps_app_status_string(collect) & "\n")
            Return collect
        End If

        While (True)
            t = Promact_isApi.ps_collect_resp(collect, length, result, -1)
            If (t = PromiraAppStatus.PS_APP_NO_MORE_CMDS_TO_COLLECT) Then
                Exit While
            ElseIf (t < 0) Then
                Console.WriteLine(Promact_isApi.ps_app_status_string(t) & "\n")
                Return t
            End If
            If (t = PromiraSpiCommand.PS_SPI_CMD_READ) Then
                If (buf.Length > offset) Then
                    Dim dataIn(buf.Length - offset) As Byte
                    ret = Promact_isApi.ps_collect_spi_read(collect, word_size, buf.Length - offset, dataIn)
                    Call Array.Copy(dataIn, 0, buf, offset, ret)
                    offset += ret
                End If
            End If
        End While

        Return offset
    End Function

    Private Sub spi_master_oe(channel As Integer, queue As Integer, enable As Byte)
        Dim collect As Integer
        Dim queue_type As Byte
        Dim buf(0) As Byte

        Call Promact_isApi.ps_queue_clear(queue)
        Call Promact_isApi.ps_queue_spi_oe(queue, enable)
        collect = Promact_isApi.ps_queue_submit(queue, channel, 0, queue_type)
        dev_collect(collect, buf)
    End Sub


    '==========================================================================
    ' MAIN PROGRAM
    '==========================================================================
    Sub spi_eeprom_run()
        Dim pm As Integer
        Dim conn As Integer
        Dim channel As Integer
        Dim queue As Integer
        Dim ip As String
        Dim res As Integer

        ' Find device
        ip = find_device_ip()
        If ip.Length = 0 Then
            Console.WriteLine("error: no available port")
            Exit Sub
        End If

        ' Open the device
        res = dev_open(ip, pm, conn, channel)
        If res < 0 Then
            Console.WriteLine("Unable to open Promira platform")
            Console.WriteLine("Error code = " & res)
            Exit Sub
        End If

        ' Ensure that the SPI subsystem is enabled
        Call Promact_isApi.ps_app_configure(channel, Promact_isApi.PS_APP_CONFIG_I2C Or Promact_isApi.PS_APP_CONFIG_SPI)

        ' Power the board using the Promira adapter's power supply.
        Call Promact_isApi.ps_phy_target_power(channel, Promact_isApi.PS_PHY_TARGET_POWER_BOTH)

        ' Setup the clock phase.
        Call Promact_isApi.ps_spi_configure(channel, PromiraSpiMode.PS_SPI_MODE_0, PromiraSpiBitorder.PS_SPI_BITORDER_MSB, 0)

        ' Configure SS
        Call Promact_isApi.ps_spi_enable_ss(channel, &H1)

        ' Set the bitrate
        Dim bitrate As Long
        bitrate = Promact_isApi.ps_spi_bitrate(channel, SPI_BITRATE)
        Console.WriteLine("Bitrate set to " & bitrate & " kHz")

        ' Create a queue for SPI transactions
        queue = Promact_isApi.ps_queue_create(conn, Promact_isApi.PS_MODULE_ID_SPI_ACTIVE)

        ' Enable master output
        Call spi_master_oe(channel, queue, 1)

        ' Write the offset and read the data
        Dim data_out(3) As Byte
        Dim data_in(18) As Byte
        Dim result As Long
        Dim collect As Integer
        Dim queue_type As Byte

        ' Set read command and address
        data_out(0) = &H3
        data_out(1) = 0
        data_out(2) = 0

        ' Write the transaction
        Call Promact_isApi.ps_queue_spi_ss(queue, &H1)
        Call Promact_isApi.ps_queue_spi_write(queue, PromiraSpiIOMode.PS_SPI_IO_STANDARD,
                                              8, 3, data_out)
        Call Promact_isApi.ps_queue_spi_write_word(queue, PromiraSpiIOMode.PS_SPI_IO_STANDARD,
                                                   8, 16, 0)
        Call Promact_isApi.ps_queue_spi_ss(queue, &H0)

        collect = Promact_isApi.ps_queue_submit(queue, channel, 0, queue_type)
        result = dev_collect(collect, data_in)

        If result < 0 Then
            Console.WriteLine("spi write error")
        Else
            Dim i As Integer
            Console.WriteLine("Read data bytes:")
            For i = 0 To 15
                ' First 3 bytes are command and address, so add 3
                Console.WriteLine(data_in(i + 3))
            Next
        End If

        ' Disable master output
        Call spi_master_oe(channel, queue, 0)

        ' Destroy the queue
        Call Promact_isApi.ps_queue_destroy(queue)

        ' Close the device and exit
        dev_close(pm, conn, channel)
    End Sub
End Module
